### PR TITLE
fix(generate-code)!: generate code URL issues

### LINKED ISSUE #6915: Generate code fails for requests with queries that contain percent character (%)
### I have checked the following:

- [x] I have searched existing issues and found nothing related to my issue.

### This bug is:

- [ ] making Bruno unusable for me
- [x] slowing me down but I'm able to continue working
- [x] annoying
- [ ] this feature was working in a previous version but is broken in the current release.

### Bruno version

3.0.2

### Operating System

Windows 11 23H2 22631.6345

### Describe the bug

When clicking on "Generate Code" with a request that has a query with a percent symbol that is not the beginning of an encoded character, for example those needed for SendGrid's message activity API, the functionality fails.

SendGrid's API accepts the operators `LIKE` and `NOT LIKE` which use the percentage symbol as a wildcard. In that case, a query can contain text like `to_email LIKE '%test.com'`. When that happens Bruno returns `Error generating code snippet`

### .bru file to reproduce the bug

_No response_

### Screenshots/Live demo link

Image

### PR DESCRIPTION
**Pull request by shubh-bruno**
### Description

BREAKING CHANGES for how encoding works in URL and in Code Generation

[JIRA](https://usebruno.atlassian.net/browse/BRU-3242)

Fixes : #6915, #7913, #8107, #4989, #6268, #7653, #5788, #5919

#### Bruno's URL encoding had four overlapping bugs:

- Path-param values like :id = aaa#bbb got inserted raw, silently losing data on the wire
- Pre-encoded URLs (%20) got double-encoded to `%2520` on the path side
- `#` handling was inconsistent across modes (sometimes stripped, sometimes encoded)
- Generate Code and the runtime used two separate pipelines that diverged

#### Solution
- Path-side encoding is now idempotent (decodeURIComponent → encodeURIComponent), so pre-encoded values survive without becoming `%2520`
- `#` is always treated as a data byte - ON encodes it to %23, OFF preserves the literal (axios strips per RFC on wire — that's by design)
- Centralized `buildHar()` in @usebruno/common — Generate Code now calls the same HAR pipeline the runtime will, so snippet matches what gets sent
- Single-pass path-param encoding via hashPathParamPositions — places a placeholder before encodeUrl runs, so reserved chars in values can't be confused with URL structure

#### Contribution Checklist:

- [ ] **I've used AI significantly to create this pull request**
- [ ] **The pull request only addresses one issue or adds one feature.**
- [ ] **The pull request does not introduce any breaking changes**
- [ ] **I have added screenshots or gifs to help explain the change if applicable.**
- [ ] **I have read the [contribution guidelines](https://github.com/usebruno/bruno/blob/main/contributing.md).**
- [ ] **Create an issue and link to the pull request.**

Note: Keeping the PR small and focused helps make it easier to review and merge. If you have multiple changes you want to make, please consider submitting them as separate pull requests.

#### Publishing to New Package Managers

Please see [here](../publishing.md) for more information.

## Summary by CodeRabbit (updated)

- Overview
  - Implements a canonical HAR pipeline and fixes multiple URL-encoding bugs (path-param raw insertion, double-encoding of pre-encoded values, inconsistent `#` handling, and snippet/runtime divergence) by centralizing HAR construction and tightening URL encoding semantics across packages.

- packages/bruno-common
  - Added generate-code/har buildHar(): canonical pipeline converting Bruno requests → HAR, with:
    - optional interpolation vs hashing of {{vars}} (provides unhash() to restore templates)
    - idempotent path-side encoding (hash path param positions to avoid double-encoding)
    - returns rawUrl and encodedUrl, effectiveAuth, and permissive URL validation
    - builds queryString and postData for supported body modes, merges/filters headers, and resolves auth (including oauth2 credential lookup)
  - Exported types: BrunoKV, BrunoBody, BrunoAuth, BrunoRequest, OAuth2CredentialRecord, BuildHarInput/Output, HarRequest, and buildHar.
  - Reworked utils/url:
    - Added safeDecodeURIComponent and made parseQueryParams accept stripFragment
    - Rewrote encodeUrl to treat `#` as data when encode is ON (encoded as `%23`), preserve fragment behavior when encode OFF, and parse/query rebuild to avoid losing `#` in query values.
  - Extensive unit tests added for buildHar and url utilities, including regressions and encoding matrices.

- packages/bruno-app
  - Rewrote Generate Code snippet-generator to use common.buildHar (and its rawUrl/encodedUrl/unhash) instead of local HAR construction; removed local buildHarRequest implementation.
  - Changed snippet post-processing: use buildHar outputs to select encoded vs raw URLs per settings.encodeUrl, restore {{var}} placeholders (via unhash), and normalize http-space quirks; preserved curl-only auth flags handling.
  - Tests adjusted:
    - Updated snippet-generator.spec to rely on reconstructed HAR URL behavior and removed mocks for internal har/auth utilities; added regression tests to ensure generator accepts previously problematic characters.
  - Minor reducer import cleanup to use new url utils.

- tests & fixtures
  - Large suite of new automated tests and fixtures (Jest + Playwright + BRU requests) covering:
    - URL encoding ON/OFF matrices for queries, path segments, and path-parameter substitution (spaces, brackets, unicode, pre-encoded inputs, `#`, `/`, `?`, `+`, `@`, `:`, commas, arrays, OAuth callback fragments, SPA hashes).
    - Regression suites ensuring snippet generation no longer triggers HTTPSnippet HAR validation errors for spaces, brackets, angle brackets, raw unicode.
    - Playwright specs to assert send-vs-snippet server-observed URL differences and Generate Code UI behavior (encode toggle).
  - tests/utils page helpers added for interacting with Generate Code UI in Playwright (getGeneratedSnippet, close dialog, openRequestInFolder, setUrlEncoding).

- packages/bruno-tests
  - Added an echo route used by Playwright tests to assert server-observed URL (returns decoded originalUrl fallback).

- Important behavioral changes / fixes
  - Path-side encoding is now content-blind idempotent (decodeURIComponent → encodeURIComponent) to avoid double-encoding pre-encoded values.
  - `#` is treated consistently: encoded to `%23` when encodeUrl is ON (so it’s preserved as data); when OFF the literal `#` is preserved in generated snippets (runtime/axios may strip as per RFC on wire).
  - Generate Code and runtime share the same HAR pipeline (buildHar) to align generated snippets with actual requests.
  - Single-pass path-parameter encoding with hashed placeholders prevents reserved characters in values from being mistaken for URL structure.

- Exports & public API surface
  - New/changed exports from common: buildHar and related types; safeDecodeURIComponent re-exported from common utils index.
  - No public API signature changes in bruno-app; generateSnippet API unchanged.

- Risk / review notes
  - High-impact, cross-package change: significant new code in packages/bruno-common (buildHar + types) and many new/updated tests covering many edge cases — review effort: High.
  - Removal of the old local har builder in bruno-app and reliance on shared buildHar requires careful review of auth/header deduplication, Content-Type defaulting, and interpolation vs hashing semantics.

- Tests & CI
  - Extensive new unit and integration tests added; many Playwright end-to-end fixtures and helper utilities included to validate encoding matrices and snippet/runtime parity.

### CODE REVIEW

File: packages/bruno-common/src/generate-code/har/index.ts [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+    case 'formUrlEncoded': {
+      const arr: any[] = Array.isArray(body.formUrlEncoded) ? body.formUrlEncoded : [];
+      const enabled = arr.filter((p) => p?.enabled);
+      const reduced = enabled.reduce<Record<string, string>>((acc, p) => {
+        acc[p.name] = p.value; return acc;
+      }, {});
+      return {
+        mimeType,
+        text: new URLSearchParams(reduced).toString(),
+        params: enabled.map((p) => ({ name: p.name, value: p.value }))
```
[the remark this answers is not in the document]
Comment (sid-bruno) [reply packages/bruno-common/src/generate-code/har/index.ts]: @shubh-bruno

File: packages/bruno-app/src/utils/url/index.js [hunk 2]
```diff
@@ -111,6 +111,8 @@ export const interpolateUrl = ({ url, variables }) => {
 };
 
 export const interpolateUrlPathParams = (url, params, variables = {}, options = {}) => {
+  const substituteValue = (value) => options.encodeUrl ? encodeURIComponent(value) : value;
```
[the remark this answers is not in the document]
Comment (sid-bruno) [reply packages/bruno-app/src/utils/url/index.js]: @shubh-bruno

File: packages/bruno-app/src/components/Sidebar/Collections/Collection/CollectionItem/GenerateCodeItem/utils/snippet-generator.js [hunk 3]
```diff
@@ -46,65 +45,59 @@ const generateSnippet = ({ language, item, collection, shouldInterpolate = false
...
+      request: {
+        method: request.method,
+        url: sourceUrl,
+        params: request.params,
+        pathParams: [],
+        headers: mergedHeaders,
+        body: request.body,
+        auth: effectiveAuth,
+        settings
+      },
```
[the remark this answers is not in the document]
Comment (sid-bruno) [reply packages/bruno-app/src/components/Sidebar/Collections/Collection/CollectionItem/GenerateCodeItem/utils/snippet-generator.js:64]: @shubh-bruno

File: tests/request/generate-code/send-vs-snippet.spec.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    await sendRequest(page, 200);
+    await expectHttpbinReceived(page, 'https://httpbin.org/anything/docs/api');
+  });
+
+  test('Scenario 1 (ON variant): page anchor — # encoded as data reaches server', async ({ pageWithUserData: page }) => {
+    await openCollection(page, COLLECTION);
+    await openRequestInFolder(page, FOLDER, 'docs-fragment-external');
+    await setUrlEncoding(page, true);
+    await sendRequest(page, 200);
+    await expectHttpbinReceived(page, 'https://httpbin.org/anything/docs/api#authentication');
```
[the remark this answers is not in the document]
Comment (shubh-bruno) [reply tests/request/generate-code/send-vs-snippet.spec.ts]: @coderabbitai `httpbin URL-decodes %23 back to # for display` httpbin.org's `/anything` endpoint URL-decodes the path when populating the "url" field in its JSON response. That's a httpbin-side behavior — Bruno did its job correctly (sending %23 on the wire), but httpbin shows us the decoded form.

**Comment by sid-bruno** [comment]
LGTM, rest is on murphy's law.

**Review by sid-bruno — APPROVED** [review]

**Comment by crowjake** [comment]
Sorry if this isn't the right place, but I see this was fixed in the nightlies a month or so ago... how long does it take for something like this to make it into release versions?

I see there's 15/17 tests passing, are those two failing tests the blockers?

**Comment by shubh-bruno** [comment]
@crowjake The fix is available in the latest **v4.0.0** release.

If you happen to come across any issues with **Generate Code** in v4, we'd appreciate it if you could open a separate issue with the relevant details so we can look into it.
